### ISSUE TITLE
"Previous Fabric installation detected on machine" - cannot find any cluster artifacts at all for standalone on-prem installation

### ISSUE THREAD

**Issue by Ryanman** [microsoft/service-fabric#967]
This is a combo issue and feature request I assume. These scripts need to show where previous fabric installations are detected, I'm at my wit's end. Version 7.1.417.9590 .

This message is profoundly not helpful when in combination with CleanFabric.ps1 no longer showing what it's doing or should be doing. Scenario/repro (does not include restarts):

1. Created a cluster from the SDK download (Unsecured, 1 node)
2. Removed this cluster via local cluster manager
3. Uninstalled SDK, uninstalled runtime
4. Downloaded Standalone package, set up on a secondary drive (E)
5. Installed SDK
6. Setup Standalone Cluster with Windows Auth, on E. Encountered issues when deploying/running.
7. Ran CleanFabric, with some folders on C: being locked and the folders on E being deleted, services stopped
8. Attempted to setup cluster, complained of previous installation. Uninstalled and reinstalled SDK.
9. Manually deleted every single SF folder I could find besides the SDK, to no avail at all.
10. Attempted to setup the cluster with the -force flag, with it erroring out on "Best practices" and previous fabric installation.

Areas checked in both C and E:
- C:/Program Data
- C:/Program Files/Microsoft SDKs (SDK is installed here)
- No services running
- Every folder in HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Service Fabric
- Every folder in HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Service Fabric SDK

Note - the fact that there is a Service Fabric Registry key is puzzling. The date it says that the last boot up was appears to be when I ran that script.

If these powershell scripts were on GH I'd dig into it myself to help figure out what conventions are used here for clean and test. 

![Test](https://user-images.githubusercontent.com/1376624/84902652-1485cb00-b07b-11ea-9e0e-0b9ba070895a.png)

I can provide any other additional information and will update this issue if I solve it on my own.

One other minor note, the contributing guidelines are out of date, as they specify creating issues in

**Comment by Ryanman** [comment]
For posterity - This issue was "Solved" by manually renaming the reg key  
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Service Fabric 
and therefore soft-deleting it.

To be honest those at MS probably are accustomed to how SF works, but the way this product behaves with atypical networking and various non-Azure scenarios is atrocious. I imagine other MS frameworks like Kubernetes have problems too but at least there's a community to rubber duck off of.

I'm not closing this issue because:

- CleanFabric should be deleting these registry keys, full stop. The failure of the script to do so or inform me that it's an action that should be taken is a simple QoL / bug fix that should have been in before V7.
- CleanFabric's activities should be better documented in general. Yes getting this to OpenSource will help people like me who don't mind digging in the source code, but not everyone (even every developer) is going to be able to do that.

**Comment by maburlik** [comment]
You're right CleanFabric.ps1 should be cleaning for this case, and we have implemented the mitigation for this exact case in the runtime installer exe which is run as part of the SDK WebPI installation - in the upgrade as well as in the uninstall path. What it does is removes the install hooks for the product - the Add-Remove-Programs entry and the registry key.

If you want access to this automitigation in case you hit the issue again, here is the link to the release notes for the latest release:
https://github.com/microsoft/service-fabric/blob/master/release_notes/Service-Fabric-71CU1-releasenotes.md#repositories-and-download-links

which list the link to the runtime exe:
Windows Developer Set-up
https://download.microsoft.com/download/a/4/e/a4e59def-7e11-4a00-b43a-4f769d89129b/MicrosoftServiceFabric.7.1.417.9590.exe
the command is
MicrosoftServiceFabric.7.1.417.9590.exe /uninstall

**Comment by Ryanman** [comment]
@maburlik  Let me talk back to make sure I understand correctly: The Uninstall for the SDK is more complete than the CleanFabric.ps1 available in the Standalone package, and is what I should lean on instead.

Is an update to CleanFabric (to be more complete) on the docket at all? I'm considering those who are deploying with the Standalone package because they may not have access to the Web PI, which is a subset of the SDK.

In any case I truly appreciate the response.

**Comment by maburlik** [comment]
When the uninstall mitigation had been added to the SDK runtime installer we were only aware this had been an issue for the SDK setup scenario and it appears this was the same case here. There were dev scenario-related handles open against the installation which were responsible for the partial uninstall, but for different scenarios it's possible the uninstallation may be blocked at different stages for which we would need logs to identify. The mitigation that was implemented will be run in an SDK upgrade scenario to automatically unblock the installation of the next version, thereby cleaning the machine of the wedge case.

FYI the implementation detail of the hook used by BPA for FabricInstallable is the presence of the registry key and/or FabricHostSvc on the destination machine. The automitigation only covers the former case and not the latter. If the service has been left behind on the machine due to a blocked/partial uninstallation you can run:
sc stop FabricHostSvc
sc query FabricHostSvc (to confirm stop)
sc delete FabricHostSvc
